### PR TITLE
MRG: Allow preload=False in read_epochs

### LINKED ISSUE #2345: Add event_id to read_epochs or selectively read
If you only want to load a subset of your epochs from disk, suppplying an event_id could save on memory. currently, it loads all the epochs, and then you can do a selection. for long experiments, and large numbers of epochs, this is not ideal. the event id would find the indices in the events and from there, it can index and load just those epochs. 

one thing that I am not certain of is whether all the epochs would still be temporarily stored in memory. if that's the case, is there any way to read only select epochs?

cc: @Eric89GXL

### PR DESCRIPTION
**Pull request by Eric89GXL**
Closes #2345.

@teonlamont this should work for you.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
Ready for review/merge from my end.

**Comment by agramfort** [comment]
before I look in details. Just to follow. Now if I have 1GB of -epo.fif I can load just one epoch without allocating 1GB or memory even during IO?

**Comment by Eric89GXL** [comment]
Yep. You can also do `epochs_a = read_epochs(..., preload=False)['a']` to only read condition `'a'` to solve @teonlamont's use case.

**Comment by agramfort** [comment]
so the data are not stored as one 3d float32* in the fif file anymore?

**Comment by Eric89GXL** [comment]
They are, nothing has changed about how the data are stored. I'll point it out.

File: mne/epochs.py [hunk 1]
```diff
@@ -2201,25 +2207,139 @@ def read_epochs(fname, proj=True, add_eeg_ref=True, verbose=None):
...
+        # >>> data = data[idx]
+        #
+        # Eventually this could be refactored in io/tag.py if other functions
+        # could make use of it
+
+        raw.fid.seek(raw.data_tag.pos + offset + 16, 0)  # 16 = Tag header
+        data = np.fromstring(raw.fid.read(size), '>f4').astype(np.float64)
+        data.shape = raw.epoch_shape
+        data *= raw.cals
+        return data
```
Comment (Eric89GXL) [inline mne/epochs.py]: @agramfort look here for the on-demand reading bit

Comment (agramfort) [inline mne/epochs.py]: ok thanks.

**Comment by agramfort** [comment]
@teonlamont give it a try to see how it fixes your problem.

File: mne/epochs.py [hunk 2]
```diff
@@ -2201,25 +2207,139 @@ def read_epochs(fname, proj=True, add_eeg_ref=True, verbose=None):
...
+    def __init__(self, fid, data_tag, event_samps, epoch_shape, cals):
+        self.fid = fid
+        self.data_tag = data_tag
+        self.event_samps = event_samps
+        self.epoch_shape = epoch_shape
+        self.cals = cals
+        self.proj = False
+
+
+class EpochsFIF(_BaseEpochs):
```
Comment (agramfort) [inline mne/epochs.py]: why a new class?

Comment (Eric89GXL) [inline mne/epochs.py]: Need to provide the `_get_epoch_from_raw(self, idx, verbose=None)` method. FWIW it actually makes more sense to me to have a different class on reading `-epo.fif` than creation of an `Epochs(raw, ...)` object. They provide the same data and the same interfaces, but how they operate under the hood is a bit different, which is how our other IO-style classes are (including `EpochsArray` and `EpochsKIT`).

Comment (agramfort) [inline mne/epochs.py]: ok sounds good

@teonlamont please give it a try. You asked for it.

**Comment by teonlamont** [comment]
@Eric89GXL although it doesn't preload the data like it should, it still tries to apply baseline parameters to each event.
```python
In [45]: epochs = mne.read_epochs(fname_epo, preload=False)
Reading /Volumes/office/Experiments/E-MEG/data/A0023/mne/A0023_OLDT_coreg_calm_iir_filt-epo.fif ...
    Found the data of interest:
        t =    -200.00 ...     600.00 ms (None)
        0 CTF compensation matrices available
858 matching events found
858 matching events found
0 projection items activated
No baseline correction applied...
No baseline correction applied...
No baseline correction applied...
No baseline correction applied...
...
0 bad epochs dropped

In [46]: epochs
Out[46]: 
<EpochsFIF  |  n_events : 858 (all good), tmin : -0.2 (s), tmax : 0.6 (s), baseline : None,
 u'33': 48, u'34': 47, u'37': 96, u'41': 191, u'42': 190, u'45': 48, u'46': 48, u'49': 48, u'50': 46, u'61': 48, u'62': 48>

In [47]: epochs._data

In [48]: 
```

**Comment by Eric89GXL** [comment]
Is that a problem?

**Comment by teonlamont** [comment]
yeah, b/c it seems like it's loading the data temporarily and just not saving it, instead of just loading the header info and such.
I think it should be the case where it only does the preprocessing of the epochs (baseline, detrend, etc.) when the data are called.

**Comment by Eric89GXL** [comment]
Ahh, I see what you mean. I think there is a call to `drop_bad_epochs` that doesn't need to be there. I can fix that hopefully soon.

**Comment by Eric89GXL** [comment]
@teonlamont can you try again? Comment has been addressed and the CIs are happy.

**Comment by teonlamont** [comment]
it works beautifully :) thanks @Eric89GXL!

**Comment by agramfort** [comment]
Feel free to merge then.  Update what's new if not done yet

**Comment by Eric89GXL** [comment]
whats_new.py updated, feel free to merge

**Comment by agramfort** [comment]
Thanks

**Pull request merged**